archive/miu_language/decision_suf.lean

```diff
@@ -0,0 +1,382 @@
+  : derivable z :=
+begin
+  induction m with k hk,
+  { revert h,
+    simp only [list.repeat, zero_mul, append_nil, imp_self], },
+  apply hk,
+  simp only [succ_mul,repeat_add] at h,
+  change repeat U 2 with [U,U] at h,
+  rw ←(append_nil (z ++ repeat U (k*2) )),
+  apply derivable.r4,
+  simp only [append_nil,append_assoc,h],
+end
+
+
+/-!
+In fine-tuning my application of `simp`, I issued the following commend to determine which lemmas
+`simp` uses.
+
+`set_option trace.simplify.rewrite true`
+-/
+
+
+/--
+We may replace several consecutive occurrences of  `"III"` with the same number of `"U"`s.
+In application of the following lemma, `xs` will either be `[]` or `[U]`.
+-/
+lemma der_cons_repeat_I_repeat_U_append_of_der_cons_repeat_I_append (c k : ℕ)
+  (hc : c % 3 = 1 ∨ c % 3 = 2) (xs : miustr) (hder : derivable (M ::(repeat I (c+3*k)) ++ xs)) :
+    derivable (M::(repeat I c ++ repeat U k) ++ xs) :=
+begin
+  revert xs,
+  induction k with a ha,
+  { simp only [list.repeat,mul_zero,add_zero,append_nil, forall_true_iff,imp_self],},
+  intro xs,
+  specialize ha (U::xs),
+  intro h₂,
+  simp only [succ_eq_add_one,repeat_add], -- We massage the goal
+  rw [←append_assoc,←cons_append],        -- into a form amenable
+  change repeat U 1 with [U],             -- to the application of
+  rw [append_assoc,singleton_append],     -- ha.
+  apply ha,
+  apply derivable.r3,
+  change [I,I,I] with repeat I 3,
+  simp only [cons_append,←repeat_add],
+  convert h₂,
+end
+
+
+/-!
+### Arithmetic
+
+We collect purely arithmetic lemmas: `add_mod2` is used to ensure we have an even number of `U`s
+while `le_pow2_and_pow2_eq_mod3` treats the congruence condition modulo 3.
+-/
+section arithmetic
+
+/--
+For every `a`, the number `a + a % 2` is even.
+-/
+lemma add_mod2 (a : ℕ) : ∃ t, a + a % 2 = t*2 :=
+begin
+  simp only [mul_comm _ 2], -- write `t*2` as `2*t`
+  apply dvd_of_mod_eq_zero, -- it suffices to prove `(a + a % 2) % 2 = 0`
+  rw [add_mod,mod_mod,←two_mul,mul_mod_right],
+end
+
+private lemma le_pow2_and_pow2_eq_mod3' (c : ℕ) (x : ℕ) (h : c = 1 ∨ c = 2) :
+  ∃ m : ℕ, c + 3*x ≤ 2^m ∧ 2^m % 3 = c % 3 :=
+begin
+  induction x with k hk,
+  { use (c+1),
+    cases h with hc hc;
+    { rw hc, norm_num }, },
+  rcases hk with ⟨g, hkg, hgmod⟩,
+  by_cases hp : (c + 3*(k+1) ≤ 2 ^g),
+  { use g, exact ⟨hp,hgmod⟩ },
+  use (g+2),
+  split,
+  { rw [mul_succ, ←add_assoc,nat.pow_add],
+    change 2^2 with (1+3), rw [mul_add (2^g) 1 3, mul_one],
+    linarith [hkg,one_le_two_pow g], },
+  rw [nat.pow_add,←mul_one c],
+  exact modeq.modeq_mul hgmod rfl,
+end
+
+/--
+If `a` is 1 or 2 modulo 3, then exists `k` a power of 2 for which `a ≤ k` and `a ≡ k [MOD 3]`.
+-/
+lemma le_pow2_and_pow2_eq_mod3 (a : ℕ) (h : a % 3 = 1 ∨ a % 3 = 2) :
+  ∃ m : ℕ, a ≤ 2^m ∧ 2^m % 3 = a % 3:=
+begin
+  cases le_pow2_and_pow2_eq_mod3' (a%3) (a/3) h with m hm,
+  use m,
+  split,
+  { convert hm.1, exact (mod_add_div a 3).symm, },
+  rw [hm.2, mod_mod _ 3],
```

Comment:
```suggestion
  { rw [hm.2, mod_mod _ 3], }
```
I'm seeing this a lot in the PR. Personally I don't like the asymmetry; `split` creates two goals, so both should be in `{ }`, especially when both branches are of similar length/difficulty. But I'm not sure we have a written standard about this, do others agree?

Reply:
I've made the changes you suggested, ensuring each remaining subgoal is in a brace-delimited indented block.

The format I used was one of the options suggested in TPIL (Section 5.4):
> [...] every time a tactic leaves more than one subgoal, we separate the remaining subgoals by enclosing them in blocks and indenting, until we are back down to one subgoal. Thus if the application of theorem foo to a single goal produces four subgoals, one would expect the proof to look like this:
> ```lean
> begin
>   apply foo,
>   { ... proof of first goal ... },
>   { ... proof of second goal ... },
>   { ... proof of third goal ... },
>   proof of final goal
> end
> ```

Do let me know if either (1) I've interpreted TPIL incorrectly or (2) the suggestions in TPIL are not sufficiently prescriptive for contributions to mathlib.

Reply:
No, you've read TPIL right! But I disagree with TPIL (except when the final goal is significantly longer than the first branch). And it looks like the [mathlib style guide](https://leanprover-community.github.io/contribute/style.html#tactic-mode) does too. We'll try to fix the discrepancy. Thanks for making the change.

Besides style, delimiting every subgoal is slightly more stable. If `foo` were changed so that `apply foo` created 5 goals instead of 4, the final proof could go haywire; the breakage is easier to diagnose when everything is focused. This is admittedly a minor point.